src/lib/error/api-error.ts

```diff
@@ -0,0 +1,270 @@
+            return 400;
+        case 'BadDataError':
+            return 400;
+        case 'BadRequestError':
+            return 400;
+        case 'OwaspValidationError':
+            return 400;
+        case 'PasswordUndefinedError':
+            return 400;
+        case 'MinimumOneEnvironmentError':
+            return 400;
+        case 'InvalidTokenError':
+            return 401;
+        case 'NoAccessError':
+            return 403;
+        case 'UsedTokenError':
+            return 403;
+        case 'InvalidOperationError':
+            return 403;
+        case 'IncompatibleProjectError':
+            return 403;
+        case 'OperationDeniedError':
+            return 403;
+        case 'NotFoundError':
+            return 404;
+        case 'NameExistsError':
+            return 409;
+        case 'FeatureHasTagError':
+            return 409;
+        case 'RoleInUseError':
+            return 400;
+        case 'ProjectWithoutOwnerError':
+            return 409;
+        case 'UnknownError':
+            return 500;
+        case 'InternalError':
+            return 500;
+        case 'PasswordMismatch':
+        case 'PasswordMismatchError':
+            return 401;
+        case 'DisabledError':
+            return 422;
+        case 'NotImplementedError':
+            return 405;
+        case 'NoAccessError':
+            return 403;
+        case 'AuthenticationRequired':
+            return 401;
+    }
+};
+
+type ValidationErrorDescription = { description: string; path?: string };
+
+type UnleashErrorData =
+    | {
+          message: string;
+          documentationLink?: string;
+      } & (
+          | {
+                name: Exclude<
+                    UnleashApiErrorKind,
+                    | 'NoAccessError'
+                    | 'AuthenticationRequired'
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                >;
+            }
+          | {
+                name: 'NoAccessError';
+                permission: string;
+            }
+          | {
+                name: 'AuthenticationRequired';
+                path: string;
+                type: string;
+            }
+          | {
+                name:
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                    | 'MinimumOneEnvironmentError';
+                errors: [
+                    ValidationErrorDescription,
+                    ...ValidationErrorDescription[],
+                ];
+            }
+      );
+
+export class UnleashError extends Error {
+    id: string;
+
+    name: UnleashApiErrorKind;
+
+    documentationLink: string | null;
+
+    statusCode: number;
+
+    additionalParameters: object;
+
+    constructor({
+        name,
+        message,
+        documentationLink,
+        ...rest
+    }: UnleashErrorData) {
+        super();
+        this.id = uuidV4();
```

Comment:
it's convenient and useful but not testable since we're baking randomness without a way to control it. But maybe it's a tradeoff worth taking.

Reply:
Interesting point! I wouldn't mind taking this out and doing it in some other way, but maybe save that for a later PR?

Reply:
Sure, that was a non-blocking comment about design principles (testability vs convenience)